### ISSUE TITLE
Adding New Method

### ISSUE THREAD

**Issue by MelvinWoodley** [mlr-org/mlr#1860]
Hi,

I have started using MLR in competition and I would really like if you could add lightgbm for regression to the models that are used. I tried to start the process on my own but I was not able to do it. Lightgbm is the newest competitor to xgboost and it has lots of popularity recently in many competitions performing well.

**Comment by larskotthoff** [comment]
Have you seen the part of the [tutorial](https://mlr-org.github.io/mlr-tutorial/devel/html/create_learner/index.html) that explains how to integrate another learner?

**Comment by MelvinWoodley** [comment]
Yes, I have seen that part of the tutorial and I have created the new learner and worked through many of the issues. However when I try to use the new learner to do model tuning, the mse test errors returned are always NA's and I am not sure why.

**Comment by larskotthoff** [comment]
Well, without seeing what you have we can't help. You could open a pull request with what you have and we could discuss it there.

**Comment by MelvinWoodley** [comment]
I have absolutely no idea how pull requests work. I am fairly new to GitHub. However, this is what I have so far. I hope this is helpful. Everything seems to work fine but when I try to do parameter tuning it returns NA's for mse.test.




library(lightgbm)
library(mlr)

makeRLearner.regr.lightgbm = function() {
  makeRLearnerRegr(
    cl = "regr.lightgbm",
    package = "lightgbm",
    par.set = makeParamSet(
      
      makeIntegerLearnerParam(id ="num_iterations", default=100,lower=1),
      makeIntegerLearnerParam(id ="verbose",default=1),
      makeDiscreteLearnerParam(id = "boosting", default = "gbdt", values = c("gbdt", "dart","goss")), 
      makeNumericLearnerParam(id = "learning_rate", default = 0.1, lower = 0), 
      makeIntegerLearnerParam(id = "max_depth", default = -1, lower = -1),  
      makeIntegerLearnerParam(id = "min_data_in_leaf", default = 20, lower = 0), 
      makeIntegerLearnerParam(id = "num_leaves", default=31, lower=1),
      makeNumericLearnerParam(id = "feature_fraction", default = 1, lower = 0, upper = 1), 
      makeNumericLearnerParam(id = "bagging_fraction", default = 1, lower = 0, upper = 1),
      makeNumericLearnerParam(id = "bagging_freq", default = 0, lower = 0), 
      makeNumericLearnerParam(id = "min_gain_to_split", default = 0, lower = 0),
      makeLogicalLearnerParam(id="use_missing",default=TRUE,tunable = FALSE),
      makeNumericLearnerParam(id = "min_sum_hessian", default=10)
      
    ),
    par.vals = list(objective="regression"),
    properties = c("numerics", "weights","missings"),
    name = "LightGBM",
    short.name = "lightgbm",
    note = "First try at this"
  )
}


trainLearner.regr.lightgbm = function(.learner, .task, .subset, .weights = NULL,  ...) {
  f = getTaskDesc(.task)
  data = getTaskData(.task, .subset,target.extra = TRUE)
  lgb.data = lgb.Dataset(as.matrix(data$data), label = data$target)
  lightgbm::lgb.train(data = lgb.data,objective="regression",
                      verbosity = -1, 
                      verbose = -1,
                      record = TRUE,...)
  
}



predictLearner.regr.lightgbm= function(.learner, .model, .newdata, ...) {
  predict(.model$learner.model, as.matrix(.newdata))
}


registerS3method("makeRLearner", "regr.lightgbm", makeRLearner.regr.lightgbm)
registerS3method("trainLearner", "regr.lightgbm", trainLearner.regr.lightgbm)
registerS3method("predictLearner", "regr.lightgbm", predictLearner.regr.lightgbm)

**Comment by MelvinWoodley** [comment]
Here is a reproducible example of NA's being produced.


data(BostonHousing, package = "mlbench")
sub<-subset(BostonHousing,select=-c(chas))  #removing this variable since lightgbm can only handle numerical variables
regr.task = makeRegrTask(id = "bh", data = sub, target = "medv")


params <- makeParamSet(
  makeNumericParam("bagging_fraction",lower = 0.5,upper = 1),
  makeNumericParam("learning_rate",lower=0.1,upper=10),
  makeIntegerParam("num_iterations",lower=80L,upper=120L),
  makeDiscreteParam("boosting",values = c("gbdt","dart")),
  makeIntegerParam("max_depth",lower = 1L,upper = 10L),
  makeNumericParam("min_data_in_leaf",lower = 10L,upper = 500L),
  #makeNumericParam("subsample",lower = 0.1,upper = 1),
  makeNumericParam("bagging_freq",lower = 0,upper = 50))


ctrl = makeTuneControlIrace(maxExperiments = 200L)
rdesc = makeResampleDesc("Holdout")
res = tuneParams("regr.lightgbm", regr.task, rdesc, par.set = params, control =ctrl, show.info = FALSE)

print(res)
# Tune result:
#   Op. pars: bagging_fraction=0.672; learning_rate=8.7; num_iterations=96; boosting=gbdt; max_depth=4; min_data_in_leaf=496; bagging_freq=4.54
#   mse.test.mean=  NA

**Comment by mb706** [comment]
What helps in these cases is asking the optimization result what error messages occur:
```R
> getOptPathErrorMessages(res$opt.path)
[1] "Error in setHyperPars2.Learner(learner, insert(par.vals, args)) : \n  138.3 is not feasible for parameter 'min_data_in_leaf'!\n"
[...]
```
The solution: `min_data_in_leaf` is `Integer`, not `Numeric`, so your `params` must reflect that. If you do instead

```R
params <- makeParamSet(
  makeNumericParam("bagging_fraction",lower = 0.5,upper = 1),
  makeNumericParam("learning_rate",lower=0.1,upper=10),
  makeIntegerParam("num_iterations",lower=80L,upper=120L),
  makeDiscreteParam("boosting",values = c("gbdt","dart")),
  makeIntegerParam("max_depth",lower = 1L,upper = 10L),
  makeIntegerParam("min_data_in_leaf",lower = 10L,upper = 500L),  # !!!
  #makeNumericParam("subsample",lower = 0.1,upper = 1),
  makeNumericParam("bagging_freq",lower = 0,upper = 50))
```
you get
```R
> print(res)
Tune result:
Op. pars: bagging_fraction=0.943; learning_rate=1.85; num_iterations=81; boosting=gbdt; max_depth=5; min_data_in_leaf=42; bagging_freq=25.9
mse.test.mean=37.4
```

**Comment by MelvinWoodley** [comment]
Thank you so much for this. This is very useful and fixes the problem that I had in my code.

I have some other questions:
1. If I go through the process of making the changes to include all the parameters of lightgbm in the sample code, what then would I need to do to get you to include it as a base learner in the package?

2. I would also like to know how to handle some special parameters that are not tunable and have special use:
num_threads -- number of real CPU's
bagging_seed -- internal seed in the learner
max_bin -- a parameter to optimize data storage
histogram_pool_size -- Allows to set a maximum for the RAM used
early_stopping_round -- Early stopping to avoid overfitting.

**Comment by MelvinWoodley** [comment]
Here is a link that gives detailed description of all the parameters and their respective counterpart in xgboost:

https://sites.google.com/view/lauraepp/parameters

**Comment by larskotthoff** [comment]
1. Create a pull request (see the [github help](https://help.github.com/articles/creating-a-pull-request/)). Also please have a look at the [coding guidelines](https://github.com/mlr-org/mlr/wiki/mlr-Coding-Guidelines).

2. set `tunable = FALSE` for those.

**Comment by MelvinWoodley** [comment]
Thank you. I appreciate your help.

**Comment by Stevo15025** [comment]
@MelvinWoodley another option for allowing lightgbm to use mlr is to make a pull request on lightgbm that includes the functions mlr would use. Similar to how [liquidsvm](https://cran.r-project.org/web/packages/liquidSVM/liquidSVM.pdf) brings in their own train and predict functions. This may be better as Bernd has mentioned previously that the mlr team is not sure how to handle packages that are not available on CRAN.

**Comment by sugs01** [comment]
@MelvinWoodley I have write a multiclass script in lightgbm with mlr package. Please look at the https://zhuanlan.zhihu.com/p/31426376 

But the comment is Chinese, I hope you will not have a reading disorder. If you have any question, please contact me.

**Comment by lorenzwalthert** [comment]
What's the status of this? I could not find the PR linked [above](https://github.com/mlr-org/mlr/issues/1860#issuecomment-310459556) and I'd also like to use lightgbm with mlr. I am not sure if I understand the concern of @SteveBronder about the package being not available on CRAN. At least as of this writing, it is. I think the algorithm is quite popular on kaggle so it would be nice to have it implemented under mlr core.

**Comment by ja-thomas** [comment]
Here is a mlr learner for lightgbm:

https://github.com/ja-thomas/autoxgboost/blob/boosting_backends/R/RLearner_classif_lightgbm.R
https://github.com/ja-thomas/autoxgboost/blob/boosting_backends/R/RLearner_regr_lightgbm.R

Maybe I should put them into mlr directly or create a `mlrExtraLearner` repository for learner that are currently not perfectly tested or hard to include for testing reasons. Might also be useful for mxnet

**Comment by raitodota** [comment]
@ja-thomas 

Small correction to your **RLearner_regr_lightgbm.R** script.

In line:
pv$data = lightgbm::lgb.Dataset(data.matrix(prep$data), **label = as.numeric(train$target) - 1**, categorical_feature = feat.cols)

we shouldn't subtract 1 from label in reggresion task in order to get correct results.
